File: tracker_parser.py
"""
Parser for minimal tracker format
Converts text tracker format to structured representation
"""

import re
from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass


@dataclass
class Note:
    """Represents a single note with pitch and velocity"""
    pitch: int  # MIDI note number (0-127) or pitch name (e.g., "C4")
    velocity: int  # MIDI velocity (0-127)


@dataclass
class TrackerStep:
    """Represents one time step in the tracker"""
    notes: List[Note]  # Empty list = rest
    is_rest: bool


@dataclass
class InstrumentTrack:
    """Represents a complete track for one instrument"""
    instrument: str
    steps: List[TrackerStep]


class TrackerParser:
    """Parses minimal tracker format into structured data"""

    # Note name to MIDI number mapping
    # Includes enharmonic equivalents (Cb=B, Fb=E, E#=F, B#=C)
    NOTE_MAP = {
        'C': 0, 'C#': 1, 'Db': 1, 'D': 2, 'D#': 3, 'Eb': 3,
        'E': 4, 'Fb': 4, 'E#': 5, 'F': 5, 'F#': 6, 'Gb': 6,
        'G': 7, 'G#': 8, 'Ab': 8, 'A': 9, 'A#': 10, 'Bb': 10,
        'B': 11, 'Cb': 11, 'B#': 0
    }

    @staticmethod
    def note_to_midi(note_name: str) -> int:
        """
        Convert note name to MIDI number
        Examples: C4 -> 60, A#3 -> 58, Gb5 -> 78, Cb4 -> 59 (B3)
        """
        match = re.match(r'^([A-G][#b]?)(-?\d+)$', note_name)
        if not match:
            raise ValueError(f"Invalid note name: {note_name}")

        note, octave = match.groups()
        octave = int(octave)

        # Handle enharmonic equivalents that cross octave boundaries
        # Cb needs to go down an octave (Cb4 = B3)
        # B# needs to go up an octave (B#3 = C4)
        if note == 'Cb':
            octave -= 1
        elif note == 'B#':
            octave += 1

        # MIDI note number = (octave + 1) * 12 + note_offset
        # Middle C (C4) = 60
        midi_num = (octave + 1) * 12 + TrackerParser.NOTE_MAP[note]

        if not 0 <= midi_num <= 127:
            raise ValueError(f"Note {note_name} out of MIDI range (0-127): {midi_num}")

        return midi_num

    @staticmethod
    def parse_note_entry(entry: str) -> List[Note]:
        """
        Parse a single note entry (can be chord)
        Examples:
          "C4:80" -> [Note(60, 80)]
          "C4:70,E4:65,G4:68" -> [Note(60, 70), Note(64, 65), Note(67, 68)]
          "C4:70,E4:65," -> [Note(60, 70), Note(64, 65)] (trailing comma ignored)
          "C4:70." -> [Note(60, 70)] (trailing period ignored)
          "." -> [] (rest)
        """
        entry = entry.strip()

        # Clean up common LLM mistakes: trailing commas, periods
        entry = entry.rstrip('.,;')
        entry = entry.strip()

        if entry == '.' or entry == '':
            return []

        notes = []
        for note_str in entry.split(','):
            note_str = note_str.strip()

            # Skip empty strings (from trailing commas)
            if not note_str:
                continue

            if ':' not in note_str:
                raise ValueError(f"Invalid note format (expected NOTE:VELOCITY): {note_str}")

            pitch_str, velocity_str = note_str.split(':', 1)

            # Clean velocity string (might have trailing junk)
            velocity_str = velocity_str.strip()
            # Extract just the digits
            velocity_digits = ''.join(c for c in velocity_str if c.isdigit())

            if not velocity_digits:
                raise ValueError(f"No valid velocity found in: {velocity_str}")

            # Convert pitch name to MIDI number
            pitch = TrackerParser.note_to_midi(pitch_str.strip())
            velocity = int(velocity_digits)

            # Clamp velocity to valid MIDI range (0-127) instead of failing
            if velocity < 0:
                print(f"  Warning: Velocity {velocity} too low, clamping to 0")
                velocity = 0
            elif velocity > 127:
                print(f"  Warning: Velocity {velocity} too high, clamping to 127")
                velocity = 127

            notes.append(Note(pitch=pitch, velocity=velocity))

        return notes

    @staticmethod
    def parse_track(instrument: str, lines: List[str]) -> InstrumentTrack:
        """
        Parse a single instrument track
        Args:
            instrument: Instrument name (BASS, DRUMS, PIANO, SAX)
            lines: List of tracker lines (one per step)
        """
        steps = []

        for line_num, line in enumerate(lines, 1):
            line = line.strip()

            # Skip empty lines
            if not line:
                continue

            # Strip line number if present (format: "1 C2:80" or "1. C2:80")
            # Match optional number (with optional period) followed by whitespace
            line = re.sub(r'^\d+\.?\s+', '', line)

            try:
                notes = TrackerParser.parse_note_entry(line)
                is_rest = len(notes) == 0
                steps.append(TrackerStep(notes=notes, is_rest=is_rest))
            except ValueError as e:
                raise ValueError(f"Error in {instrument} track, line {line_num}: {e}")

        return InstrumentTrack(instrument=instrument, steps=steps)

    @staticmethod
    def parse(tracker_text: str) -> Dict[str, InstrumentTrack]:
        """
        Parse complete tracker format text
        Returns dict mapping instrument name to InstrumentTrack

        Expected format:
        BASS
        1 C2:80
        2 .
        3 E2:75
        ...

        DRUMS
        1 C1:90,F#1:60
        ...

        Note: Line numbers are optional and will be automatically stripped if present.
        """
        lines = tracker_text.strip().split('\n')
        tracks = {}
        current_instrument = None
        current_lines = []

        for line in lines:
            line = line.strip()

            # Check if this is a section header
            if line in ['BASS', 'DRUMS', 'PIANO', 'SAX']:
                # Save previous instrument if exists
                if current_instrument and current_lines:
                    tracks[current_instrument] = TrackerParser.parse_track(
                        current_instrument, current_lines
                    )

                # Start new instrument
                current_instrument = line
                current_lines = []
            elif line:  # Non-empty line, not a header
                if current_instrument is None:
                    raise ValueError(f"Found note data before instrument header: {line}")
                current_lines.append(line)

        # Don't forget the last instrument
        if current_instrument and current_lines:
            tracks[current_instrument] = TrackerParser.parse_track(
                current_instrument, current_lines
            )

        return tracks


# Convenience function
def parse_tracker(tracker_text: str) -> Dict[str, InstrumentTrack]:
    """Parse tracker format text. Returns dict of instrument -> InstrumentTrack"""
    return TrackerParser.parse(tracker_text)

File: web/js/trackerParser.js
import { totalSteps } from './config.js';

const NOTE_MAP = {
  C: 0,
  'C#': 1,
  Db: 1,
  D: 2,
  'D#': 3,
  Eb: 3,
  E: 4,
  Fb: 4,
  'E#': 5,
  F: 5,
  'F#': 6,
  Gb: 6,
  G: 7,
  'G#': 8,
  Ab: 8,
  A: 9,
  'A#': 10,
  Bb: 10,
  B: 11,
  Cb: 11,
  'B#': 0,
};

export function noteToMidi(noteName) {
  const match = /^([A-G][#b]?)(-?\d+)$/.exec(noteName.trim());
  if (!match) {
    throw new Error(`Invalid note name: ${noteName}`);
  }
  let [, note, octaveStr] = match;
  let octave = Number.parseInt(octaveStr, 10);
  if (note === 'Cb') {
    octave -= 1;
  } else if (note === 'B#') {
    octave += 1;
  }
  const offset = NOTE_MAP[note];
  if (offset === undefined) {
    throw new Error(`Unsupported note: ${noteName}`);
  }
  const midi = (octave + 1) * 12 + offset;
  if (midi < 0 || midi > 127) {
    throw new Error(`Note ${noteName} out of MIDI range (0-127): ${midi}`);
  }
  return midi;
}

export function midiToFrequency(midi) {
  return 440 * Math.pow(2, (midi - 69) / 12);
}

function cleanEntry(entry) {
  return entry.trim().replace(/[.,;]+$/g, '').trim();
}

export function parseNoteEntry(entry) {
  const cleaned = cleanEntry(entry);
  if (!cleaned || cleaned === '.') {
    return { notes: [], isTie: false };
  }
  if (cleaned === '^') {
    return { notes: [], isTie: true };
  }
  const notes = [];
  const parts = cleaned.split(',');
  for (const rawPart of parts) {
    const part = rawPart.trim();
    if (!part) continue;
    const [pitchPart, velocityPart] = part.split(':');
    if (!velocityPart) {
      throw new Error(`Invalid note format (expected NOTE:VELOCITY): ${part}`);
    }
    const velocityDigits = (velocityPart.match(/\d+/g) || []).join('');
    if (!velocityDigits) {
      throw new Error(`No valid velocity found in: ${velocityPart}`);
    }
    let velocity = Number.parseInt(velocityDigits, 10);
    velocity = Math.min(127, Math.max(0, velocity));
    const pitch = noteToMidi(pitchPart.trim());
    notes.push({ pitch, velocity });
  }
  return { notes, isTie: false };
}

function stripLineNumber(line) {
  return line.replace(/^\s*\d+\.?\s+/, '').trim();
}

export function parseTrack(instrument, lines) {
  const steps = [];
  let lineNumber = 0;
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    lineNumber += 1;
    const body = stripLineNumber(line);
    const { notes, isTie } = parseNoteEntry(body);
    steps.push({
      notes,
      isRest: notes.length === 0 && !isTie,
      isTie,
    });
  }
  return { instrument, steps };
}

export class TrackerParser {
  static parse(trackerText) {
    const lines = trackerText
      .split(/\r?\n/)
      .map((line) => line.trimEnd());
    const instruments = ['BASS', 'DRUMS', 'PIANO', 'SAX'];
    const tracks = {};
    let currentInstrument = null;
    let currentLines = [];
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      if (instruments.includes(trimmed)) {
        if (currentInstrument && currentLines.length) {
          tracks[currentInstrument] = parseTrack(currentInstrument, currentLines);
        }
        currentInstrument = trimmed;
        currentLines = [];
      } else if (currentInstrument) {
        currentLines.push(trimmed);
      }
    }
    if (currentInstrument && currentLines.length) {
      tracks[currentInstrument] = parseTrack(currentInstrument, currentLines);
    }
    return tracks;
  }
}

export class TrackerStreamParser {
  constructor(config) {
    this.config = config;
    this.totalSteps = totalSteps(config);
    this.reset();
  }

  reset() {
    this.partialLine = '';
    this.currentInstrument = null;
    this.currentStepCounts = {
      BASS: 0,
      DRUMS: 0,
      PIANO: 0,
      SAX: 0,
    };
    this.sections = {
      BASS: [],
      DRUMS: [],
      PIANO: [],
      SAX: [],
    };
    this.rawLines = [];
  }

  appendChunk(chunk) {
    const decoded = chunk.replace(/\r/g, '');
    const lines = (this.partialLine + decoded).split('\n');
    this.partialLine = lines.pop();
    const newSteps = [];
    for (const line of lines) {
      this._processLine(line, newSteps);
    }
    return newSteps;
  }

  finalize() {
    if (this.partialLine) {
      const newSteps = [];
      this._processLine(this.partialLine, newSteps);
      this.partialLine = '';
      return newSteps;
    }
    return [];
  }

  _processLine(line, newSteps) {
    const trimmed = line.trim();
    if (!trimmed) return;
    this.rawLines.push(trimmed);
    const instruments = ['BASS', 'DRUMS', 'PIANO', 'SAX'];
    if (trimmed.startsWith('#')) {
      return;
    }
    if (instruments.includes(trimmed)) {
      this.currentInstrument = trimmed;
      return;
    }
    if (!this.currentInstrument) {
      return;
    }
    if (this.currentStepCounts[this.currentInstrument] >= this.totalSteps) {
      return;
    }
    const body = stripLineNumber(trimmed);
    let parsed;
    try {
      parsed = parseNoteEntry(body);
    } catch (error) {
      console.warn('Skipping malformed tracker line', { line: trimmed, error });
      return;
    }
    const stepIndex = this.currentStepCounts[this.currentInstrument];
    this.currentStepCounts[this.currentInstrument] += 1;
    const step = {
      notes: parsed.notes,
      isRest: parsed.notes.length === 0 && !parsed.isTie,
      isTie: parsed.isTie,
    };
    this.sections[this.currentInstrument].push({
      ...step,
      index: stepIndex,
      line: trimmed,
    });
    newSteps.push({
      instrument: this.currentInstrument,
      stepIndex,
      step,
      line: trimmed,
    });
  }

  getRawTrackerText() {
    const parts = [];
    for (const instrument of ['BASS', 'DRUMS', 'PIANO', 'SAX']) {
      const header = `${instrument}`;
      parts.push(header);
      const steps = this.sections[instrument];
      for (const { line } of steps) {
        parts.push(line);
      }
      parts.push('');
    }
    return parts.join('\n').trim();
  }
}
